cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -144,6 +144,20 @@ class AllocationInstruction extends CallInstruction {
   AllocationInstruction() { this.getStaticCallTarget() instanceof Cpp::AllocationFunction }
 }
 
+private predicate isIndirectionType(Type t) { t instanceof Indirection }
+
+private predicate hasUnspecifiedBaseType(Indirection t, Type base) {
+  base = t.getBaseType().getUnspecifiedType()
+}
+
+/**
+ * Holds if `t2` is the same type as `t1`, but after stripping away `result` number
+ * of indirections.
+ * Furthermore, specifies in `t2` been deeply stripped and typedefs has been resolved.
+ */
+private int getNumberOfIndirectionsImpl(Type t1, Type t2) =
+  shortestDistances(isIndirectionType/1, hasUnspecifiedBaseType/2)(t1, t2, result)
```

Comment:
This is O(n^2) on the length of the indirection chain, right?  I suppose chains longer than about 3 are probably very rare, and this is just what the `shortestDistances` HOP does - bringing the restriction `not t2 instanceof Indirection` in to here wouldn't actually affect what it calculates.

Reply:
Correct. We can't actually bring the `not t2 instanceof Indirection` restriction in here because arguments to HOPs must be "either variables or _s" according to the compiler.

And ineed, chains longer than 3 are properly very rare so I don't think this is an issue (and DCA does confirm that this doesn't really impact performance in any way).